src/blockstore/log.go

```diff
@@ -0,0 +1,443 @@
+	var numOfLeaves, numOfNodes, firstLeafIndex int
+	var treeOffset int64
+
+	// Fist assign a new header file.
+	newHeader = new(logHeader)
+	if newHeader == nil {
+		log.Println("Could not allocate header.")
+		return errors.New("Could not allocate header.")
+	}
+	newHeader.headerVersion = defaultHeaderVersion
+	newHeader.numOfKeys = len(kvList)
+	if lm.headLog != nil {
+		newHeader.nextLog = lm.headLog
+		newHeader.nextRoot = lm.headLog.merkleRoot
+	}
+
+	//TODO move down after completion/verification of write to
+	// prevent corruption of head.
+	lm.headLog = newHeader
+
+	// Making sure we are writing a NEW file.
+	for checkFileExists(lm.generateName()) {
+		lm.nextFD++
+	}
+	newHeader.file, err = os.Create(lm.generateName())
+	if err != nil {
+		log.Println("Could not create log file.")
+		return errors.New("Could not create log file.")
+	}
+	lm.nextFD++
+
+	// Generate the header, key index table, and merkleTree in memory.
+	headerBytes, err = generateHeader(newHeader)
+	if err != nil {
+		return errors.New("Could not generate header.")
+	}
+
+	indexBytes, err = generateIndex(kvList)
+	if err != nil {
+		return errors.New("Could not generate index table.")
+	}
+
+	numOfLeaves = 1
+	for numOfLeaves < len(kvList) {
+		numOfLeaves *= 2
+	}
+	// Item [0] will be the encrypted Root, Root will be [1],
+	// and leaves are hash of blocks
+	numOfNodes = 2 * numOfLeaves
+	firstLeafIndex = numOfLeaves
+
+	hashList = make([][]byte, numOfNodes, numOfNodes)
+	if hashList == nil {
+		log.Println("Could not allocate memory for hashList.")
+		return errors.New("Could not allocate memory for hashList.")
+	}
+
+	// Pre-allocating all the space for hash list.
+	// This also gives the empty leaves a hash of 0
+	for i := 0; i < numOfNodes; i++ {
+		hashList[i] = make([]byte, hashLength, hashLength)
+		if hashList[i] == nil {
+			log.Println("Could not allocate memory for hashList.")
+			return errors.New("Could not allocate memory for hashList.")
+		}
+	}
+
+	// Writing each part to file. The tree is not filled yet, an
+	// empty tree is written, and the correct values will be written after
+	// writing the data blocks.
+	bytesWritten, err = newHeader.file.Write(headerBytes)
+	if err != nil || bytesWritten != len(headerBytes) {
+		log.Println("Could not write the header section.")
+		return errors.New("Could not write the header section.")
+	}
+
+	bytesWritten, err = newHeader.file.Write(indexBytes)
+	if err != nil || bytesWritten != len(indexBytes) {
+		log.Println("Could not write the index section.")
+		return errors.New("Could not write the index section.")
+	}
+
+	// Saving the offset, to seek and overwrite after writing data blocks.
+	treeOffset, err = newHeader.file.Seek(0, 1)
+	for i := 0; i < numOfNodes; i++ {
+		bytesWritten, err = newHeader.file.Write(hashList[i])
+		if err != nil || bytesWritten != len(hashList[i]) {
+			log.Println("Could not write the tree section.")
+			return errors.New("Could not write the tre section.")
+		}
+	}
```

Comment:
Consider using shorter idiomatic error handling patterns, for instance:

```
if bytesWritten, err  := newHeader.file.Write(headerBytes); err != nil {
    return err //or new error containing old error
}
```

see https://blog.golang.org/error-handling-and-go for how to pare down the error handling code

Reply:
Right, they are each a couple of lines longer.
I was not sure, my thought was to keep a single line from getting too long.

There are many instances of this in the code. Do you suggest that I reformat them in the next PR?

Reply:
I don't know if gofmt will do that automatically, but usually I write longer lines and let the tools shorten them if need be.